factory/src/test/resources/good/ClassWithCopyAnnotations.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2013 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package tests;
+
+import com.google.auto.factory.AutoFactory;
+import com.google.auto.factory.CopyAnnotations;
+import com.google.errorprone.annotations.Immutable;
+import org.checkerframework.framework.qual.DefaultQualifier;
+import org.checkerframework.framework.qual.DefaultQualifiers;
+import javax.inject.Singleton;
+
+@CopyAnnotations
```

Comment:
I think the annotation should apply only to the class that has the `@AutoFactory` annotation. Otherwise it's hard to know what the eventual result will be. What would happen if `ClassWithCopyAnnotations` had its own `@CopyAnnotations` annotation, for example?

Reply:
In this example the @AutoFactory is placed at the ctor.
Do you say, you want that in this case @AutoFactory.CopyAnnotations is not allowed? Only if the @AutoFactory is placed really at the class?

Reply:
@eamonnmcmanus can you please comment on this

Reply:
My earlier comment wasn't quite right. I should have said that `@AutoFactory.CopyAnnotations` should be applied to exactly the same element as `@AutoFactory`. If `@AutoFactory` is applied to a class then `@CopyAnnotations` should be applied to the same class. If `@AutoFactory` is applied to a constructor then `@CopyAnnotations` should apply to the same constructor. So in other words I'm resisting the idea that `@CopyAnnotations` applied to a class would affect all `@AutoFactory` constructors within that class.

Reply:
@eamonnmcmanus 
Until now, I had the understanding, that `@CopyAnnotations` is always on the class, because it copies the class annotations onto the factory class.

With your statement, when it is placed onto the ctor (or multiple ctors), I (as a user) would interpret it as "copy ctor annotations to the create method".

What shall be the exact semantics and allowed constellations then, one of these?
1. If `@CopyAnnotations` is present on one or multiple of the ctors, it is like as it is on the class. The arguments must be exactly the same on each one. (redundancy)
2. `@CopyAnnotations` must be present either on none or all ctors, with all args exactly the same. (redundancy)
3. `@CopyAnnotations` must be present at most once on a ctors.
4. `@CopyAnnotations` is only allowed on class, and if present, the `@AutoFactory` is as well only allowed on the class

Or, we restrict `@CopyAnnotations` today to the class, accepting that it might be a bit confusing if the `@AutoFactory` is at some ctors and not at class.
This would allow in future the extension to allow `@CopyAnnotations` onto the ctor with the semantics to copy the ctor annotation onto the factory create method.

Reply:
I think that `@CopyAnnotations` should be applied to the same thing (class or constructor) as `@AutoFactory`. I think it should not copy class annotations unless `fromClass` has a non-default value, and in that case it should obviously copy them from that class. It doesn't bother me that you might have to have _n_ very similar `@CopyAnnotations` if you have _n_ overloaded `@AutoFactory` constructors. I'm not sure that's very common. Searching over Google's internal codebase suggests that fewer than 1% of source files that contain `@AutoFactory` contain it more than once.

Reply:
@eamonnmcmanus 
the requirement is still not clear to me. Perhaps we can have a short talk, to save time? gmail user frank.rene.benoit

1. You say "should not copy class annotations unless fromClass has a non-default value". 
I thought this only switches to _another_ copy source. 
Do you say, it is mandatory to define fromClass to make it active at all?
Then the default value is useless, right?

```java
@AutoFactory
//@AutoFactory.CopyAnnotation // <- not sufficient
@AutoFactory.CopyAnnotation(fromClass = Foo.class)
@NonNullByDefault
class Foo {}
```

2. This shall give error, due to different copy source

```java
@NonNullByDefault
class Foo {
    @AutoFactory
    @AutoFactory.CopyAnnotation(fromClass = Foo.class)
    Foo(){}

    @AutoFactory
    @AutoFactory.CopyAnnotation(fromClass = Other.class)
    Foo(int arg){}
}
```

3. This is ok, only one CopyAnnotations needed.

```java
@NonNullByDefault
class Foo {
    @AutoFactory
    @AutoFactory.CopyAnnotation(fromClass = Foo.class)
    Foo(){}

    @AutoFactory
    Foo(int arg){}
}
```

Reply:
Just as a reminder, we may end up with a different design after Google's internal API review process. I haven't started that process yet but intend to soon.

Nevertheless I think the exercise of implementing a proposal is a useful one because it is revealing problems. You have shown that it is probably a mistake to try to use the same annotation to control copying of class annotations and parameter annotations. I think it makes more sense to have separate `@AutoFactory.CopyClassAnnotations` and `@AutoFactory.CopyParameterAnnotations`. Then this PR would just be for the former. In that case it means that `@AutoFactory.CopyClassAnnotations` applies to the `@AutoFactory` class, whether or not the `@AutoFactory` annotation is applied to that class or to its constructor(s). There are a few things that are unclear, though:

1. Should it also be legal to apply `@AutoFactory.CopyClassAnnotations` to a constructor that has the `@AutoFactory` annotation? Otherwise, if it's applied to the class, it could be quite far away from the `@AutoFactory` annotation. But as you mention, there's then the issue of what happens when there is more than one `@AutoFactory` constructor.

2. Should it default to copying the annotations of the `@AutoFactory` class? The alternative is for `fromClass` not to have a default.

3. Would it be sufficiently clear if we renamed `fromClass` to `value`, so you could write `@AutoFactory.CopyAnnotations(SomeClass.class)` instead of `@AutoFactory.CopyAnnotations(fromClass = SomeClass.class)`?

I think you can make arbitrary choices for these questions, but we'll need to sort them out in the API review.

Reply:
@eamonnmcmanus can you tell, when this "API review" will take place? For me this issue is blocking to start using auto-factory in my work project.

Reply:
@eamonnmcmanus can you tell, when this "API review" will take place? For me this issue is blocking to start using auto-factory in my work project.

Reply:
If you are blocked I would suggest forking AutoFactory until we are able to get to it.